content/docs/testing-recipes.md

```diff
@@ -0,0 +1,403 @@
+---
+id: testing-recipes
+title: Test recipes 
+permalink: docs/testing-recipes.html
+---
+
+This document lists some common testing patterns for React components.
+
+- This assumes tests are running in jest. For details on setting up a testing environment, read [the Environments doc](/docs/testing-environments.html).
+- This guide uses React's [`act()`](/docs/act.html) testing helper.
+
+### Rendering {#rendering}
+
+Testing whether a component renders correctly for given props is a useful signal to the stability of a component. Consider a simple component that renders a message based on a prop.
+
+```jsx
+// hello.js
+export default function Hello(props) {
+  return props.name === undefined ? (
+    <h1>Hello, {props.name}!</h1>
+  ) : (
+    <span>Hey, stranger</span>
+  );
+}
+```
+
+We can write a test for this component:
+
+```jsx
+// hello.test.js
+import Hello from "./hello";
+import { render } from "react-dom";
+import { act } from "react-dom/test-utils";
+
+let container = null
+beforeEach(() => {
+  // setup a DOM element as a render target
+  container = document.createElement('div')
+  // attach the container to document so events works correctly
+  document.body.appendChild(container)
+})
+
+afterEach(() => {
+  // cleanup on exiting
+  container.remove();
+  container = null  
+})
+
+it("renders with or without a name", () => {
+  act(() => {
+    render(<Hello />, container);
+  });
+  expect(container.textContent).toBe("Hey, stranger");
+
+  act(() => {
+    render(<Hello name="Sophie" />, container);
+  });
+  expect(container.textContent).toBe("Hello Sophie!");
+
+  act(() => {
+    render(<Hello name="Flarnie" />, container);
+  });
+  expect(container.textContent).toBe("Hello Flarnie!");
+});
+```
+
+### Data fetching {#data-fetching}
+
+Mocking data fetching with 'fake' data prevents flaky tests, and makes them run faster.
```

Comment:
```suggestion
Instead of calling real APIs in all your tests, you can mock requests with dummy data. Mocking data fetching with 'fake' data prevents flaky tests due to an unavailable backend, and makes them run faster. Note: you may still want to run a subset of tests using an "end-to-end" framework that tells whether the whole app is working together.
```